### PR TITLE
Add adapter pattern to the storage system

### PR DESCRIPTION
**Pull request by ZePedroResende**
Why ?
The whole project is a good opportunity to experiment with different types
of storage solutions and showcase some cool design patterns in elixir.
This allows us to change the storage type with a simple env var and with
a default behaviour

How ?
Add a secrets behavior
Add config module for easy loading of config defined variables

### CODE REVIEW

File: api/config/config.exs [hunk 1]
```diff
@@ -23,10 +23,15 @@ config :logger, :console,
 # Use Jason for JSON parsing in Phoenix
 config :phoenix, :json_library, Jason
 
-# Configures Redis
-redis_url = System.get_env("REDIS_URL") || "redis://localhost:6379"
+secrets_adapter = System.get_env("SECRETS_ADAPTER") || "REDIS"
```
Comment (jfranciscosousa) [inline api/config/config.exs:26]: We should throw an error on unsupported values maybe

**Comment by jfranciscosousa** [comment]
Nice work. 

But one thing, do we really want to use other storage patterns? Right now I see us literally porting over all of this to Ecto to have richer functionality, ditching redis for once. Might be cumbersome to maintain code and tests for different apdaters. New features will also need to be replicated across existing adapters. Is it really something we want to do?

**Comment by ZePedroResende** [comment]
@jfranciscosousa 
>  Right now I see us literally porting over all of this to Ecto to have richer functionality, ditching redis for once
This could be a good entrypoint to make some benchmarks and compare some types of storage systems and in memory databases. 


> Might be cumbersome to maintain code and tests for different apdaters.

If we use adapters we dont need different tests between adapters . It should support the interface and pass a default test suit for that interface. 
If it turns out to cumbersome ill remove it , its easily done, no breaking changes here, in the worst case just copy past the most used/developed adapter in the interface file and delete the adapter folder.

**Review by jfranciscosousa — APPROVED** [review]

**Review by DavideSilva — APPROVED** [review]
Fix the conflicts and :shipit:

**Pull request closed without merging**